cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,523 @@
+	// Squash!
+	podList := make([]*apiv1.Pod, total, total+1000)
+	j := 0
+	for i := 0; i < len(pods); i++ {
+		copy(podList[j:], pods[i])
+		j += len(pods[i])
+	}
+	data.podList = podList
+}
+
+func (data *internalDeltaSnapshotData) getAllPods() ([]*apiv1.Pod, error) {
+	if data.podList == nil {
+		data.buildPodList()
+	}
+	return data.podList, nil
+
+}
+
+func (data *internalDeltaSnapshotData) getAllNodes() ([]*apiv1.Node, error) {
+	nodeInfos := data.GetNodeInfoList()
+	nodes := make([]*apiv1.Node, len(nodeInfos), len(nodeInfos))
+	for i, nodeInfo := range nodeInfos {
+		nodes[i] = nodeInfo.Node()
+	}
+	return nodes, nil
+}
+
+func (data *internalDeltaSnapshotData) fork() *internalDeltaSnapshotData {
+	forkedData := newInternalDeltaSnapshotData()
+	forkedData.baseData = data
+	return forkedData
+}
+
+func (data *internalDeltaSnapshotData) commit() *internalDeltaSnapshotData {
+	for _, node := range data.nodeInfoMap {
+		data.baseData.updateNode(node)
+	}
+	for node, _ := range data.deletedNodeInfos {
+		data.baseData.removeNode(node)
+	}
+	return data.baseData
+}
+
+// NewDeltaClusterSnapshot creates instances of DeltaClusterSnapshot.
+func NewDeltaClusterSnapshot() *DeltaClusterSnapshot {
+	snapshot := &DeltaClusterSnapshot{}
+	_ = snapshot.Clear()
+	return snapshot
+}
+
+// AddNode adds node to the snapshot.
+func (snapshot *DeltaClusterSnapshot) AddNode(node *apiv1.Node) error {
+	return snapshot.data.addNode(node)
+}
+
+// AddNodes adds nodes in batch to the snapshot.
+func (snapshot *DeltaClusterSnapshot) AddNodes(nodes []*apiv1.Node) error {
+	return snapshot.data.addNodes(nodes)
+}
+
+// AddNodeWithPods adds a node and set of pods to be scheduled to this node to the snapshot.
+func (snapshot *DeltaClusterSnapshot) AddNodeWithPods(node *apiv1.Node, pods []*apiv1.Pod) error {
+	if err := snapshot.AddNode(node); err != nil {
+		return err
+	}
+	for _, pod := range pods {
+		if err := snapshot.AddPod(pod, node.Name); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RemoveNode removes nodes (and pods scheduled to it) from the snapshot.
+func (snapshot *DeltaClusterSnapshot) RemoveNode(nodeName string) error {
+	return snapshot.data.removeNode(nodeName)
+}
+
+// AddPod adds pod to the snapshot and schedules it to given node.
+func (snapshot *DeltaClusterSnapshot) AddPod(pod *apiv1.Pod, nodeName string) error {
+	return snapshot.data.addPod(pod, nodeName)
+}
+
+// RemovePod removes pod from the snapshot.
+func (snapshot *DeltaClusterSnapshot) RemovePod(namespace string, podName string) error {
+	return snapshot.data.removePod(namespace, podName)
+}
+
+// GetAllPods returns list of all the pods in snapshot
+func (snapshot *DeltaClusterSnapshot) GetAllPods() ([]*apiv1.Pod, error) {
+	return snapshot.data.getAllPods()
+}
+
+// GetAllNodes returns list of all the nodes in snapshot
+func (snapshot *DeltaClusterSnapshot) GetAllNodes() ([]*apiv1.Node, error) {
+	return snapshot.data.getAllNodes()
+}
+
+// Time: O(1)
```

Comment:
Linter will complain. First line of comment must be `// MethodName ....`

Reply:
Done